File: BOJ/stack&queue/9012_괄호.py
import sys
n = int(sys.stdin.readline())

for _ in range(n):
    li = list(sys.stdin.readline())
    stack=[]

    dic={ ")" : "(",
        "}" : "{",
       "]" : "["
       }
    n=1
    for ch in li:
        if ch in "({[":
            stack.append(ch)
        if ch in ")}]":
            if len(stack)==0:
                print("NO")
                n=0
                break
            if dic[ch]!=stack.pop():
                print("NO")
                n=0
                break
    if n==0:
        continue
    else:
        if len(stack)!=0:
            print("NO")
            continue
    print("YES")

File: BOJ/stack&queue/9012_괄호.cpp
#include <bits/stdc++.h>
using namespace std;

int T;
string s;
int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  cin >> T;
  for (int i = 0; i < T; i++){
    stack<char> stk;
    cin >> s;
    int flag = 0;
    for (char c:s)
    {
      if (c == '(') stk.push(c);
      else{
        if (stk.size())stk.pop();
        else{
          flag = 1;
          break;
        }
      }
    }
    if (stk.size() || flag)
      cout << "NO" << '\n';
    else cout << "YES" << '\n';
  }
  return 0;
}